java/arcs/sdk/android/storage/AndroidStorageServiceEndpointManager.kt

```diff
@@ -1,20 +1,121 @@
 package arcs.sdk.android.storage
 
-import android.content.Context
-import arcs.core.storage.DirectStorageEndpointManager
-import arcs.core.storage.StoreManager
+import arcs.android.crdt.toParcelableType
+import arcs.android.storage.decodeProxyMessage
+import arcs.android.storage.service.IStorageService
+import arcs.android.storage.service.IStorageServiceCallback
+import arcs.android.storage.service.suspendForRegistrationCallback
+import arcs.android.storage.service.suspendForResultCallback
+import arcs.android.storage.toProto
+import arcs.core.common.CounterFlow
+import arcs.core.crdt.CrdtData
+import arcs.core.crdt.CrdtException
+import arcs.core.crdt.CrdtOperation
+import arcs.core.crdt.CrdtOperationAtTime
+import arcs.core.storage.ProxyCallback
+import arcs.core.storage.ProxyMessage
+import arcs.core.storage.StorageEndpoint
+import arcs.core.storage.StorageEndpointManager
+import arcs.core.storage.StoreOptions
+import arcs.core.util.TaggedLog
 import arcs.sdk.android.storage.service.ConnectionFactory
-import kotlin.coroutines.CoroutineContext
+import arcs.sdk.android.storage.service.StorageServiceConnection
+import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.launch
 
+/**
+ * An implementation of a [StorageEndpointManager] that provides channels to stores that are
+ * managed by an Android [StorageService].
+ *
+ * @param coroutineContext the [CoroutineScope] that will run callbacks from the storage service.
+ * @param connectionFactory the [ConnectionFactory] to use when connecting to the storage service.
+ */
 @ExperimentalCoroutinesApi
-fun AndroidStorageServiceEndpointManager(
-  context: Context,
-  coroutineContext: CoroutineContext,
-  connectionFactory: ConnectionFactory? = null
-
-) = DirectStorageEndpointManager(
-  StoreManager(
-    ServiceStoreFactory(context, coroutineContext, connectionFactory)
-  )
-)
+class AndroidStorageServiceEndpointManager(
+  private val coroutineScope: CoroutineScope,
+  private val connectionFactory: ConnectionFactory
+) : StorageEndpointManager {
+
+  override suspend fun <Data : CrdtData, Op : CrdtOperationAtTime, T> get(
+    storeOptions: StoreOptions,
+    callback: ProxyCallback<Data, Op, T>
+  ): StorageEndpoint<Data, Op, T> {
+    val connection = connectionFactory(storeOptions, storeOptions.type.toParcelableType())
+    val service = IStorageService.Stub.asInterface(connection.connectAsync().await())
+    val channelId = suspendForRegistrationCallback { resultCallback ->
+      service.registerCallback(
+        StorageServiceProxyCallback(coroutineScope, callback),
+        resultCallback
+      )
+    }
+    return AndroidStorageEndpoint(
+      channelId, service, connection
+    )
+  }
+
+  override suspend fun reset() {}
+}
+
+@ExperimentalCoroutinesApi
+class AndroidStorageEndpoint<Data : CrdtData, Op : CrdtOperationAtTime, T>(
```

Comment:
Some KDoc here would be good, unless it's intended that the class be private to the file (thus only visible to the manager above)..

Reply:
I think the type can be public, so I added a doc. But I made the ctor private.